src/utils/acmp.cc

```diff
@@ -236,10 +236,10 @@ static void acmp_build_binary_tree(ACMP *parser, acmp_node_t *node) {
             nodes[i] = nodes[j];
             nodes[j] = tmp;
         }
-        if (node->btree != NULL) {
-            free(node->btree);
-            node->btree = NULL;
-        }
+    if (node->btree != NULL) {
```

Comment:
I see your idea and you're right, this can be confusing.

But could you add a bit more clarification?

I suggest you to add an open `{` at the and of line [229](https://github.com/owasp-modsecurity/ModSecurity/pull/3424/files#diff-09c8944650ce24e352a8a965ff221eeeb5f788735a89614db64dd6a943fd3d48R229), and a close one before this `if (...) {` block? I think that would make the whole part more clear.

Note, that we can check the behavior through regression test against [@pm](https://github.com/owasp-modsecurity/ModSecurity/blob/5d05ba359cff56adfd0bbefed5f6cab3254f1b4d/test/test-cases/regression/operator-pm.json) and [secrule-language-test/@pm](https://github.com/owasp-modsecurity/secrules-language-tests/blob/a3d4405e5a2c90488c387e589c5534974575e35b/operators/pm.json)